packages/typescript-eslint-parser/src/analyze-scope.ts

```diff
@@ -0,0 +1,790 @@
+
+/**
+ * Define the override function of `Scope#__define` for global augmentation.
+ * @param {Function} define The original Scope#__define method.
+ * @returns {Function} The override function.
+ */
+function overrideDefine(define: any) {
+  return /* @this {Scope} */ function(this: any, node: any, definition: any) {
+    define.call(this, node, definition);
+
+    // Set `variable.eslintUsed` to tell ESLint that the variable is exported.
+    const variable = this.set.get(node.name);
+    if (variable) {
+      variable.eslintUsed = true;
+    }
+  };
+}
+
+/** The scope class for enum. */
+class EnumScope extends Scope {
+  constructor(scopeManager: any, upperScope: any, block: any) {
+    super(scopeManager, 'enum', upperScope, block, false);
+  }
+}
+
+class PatternVisitor extends OriginalPatternVisitor {
+  constructor(...args: any[]) {
+    super(...args);
+  }
+
+  Identifier(node: any) {
+    super.Identifier(node);
+    if (node.decorators) {
+      this.rightHandNodes.push(...node.decorators);
+    }
+    if (node.typeAnnotation) {
+      this.rightHandNodes.push(node.typeAnnotation);
+    }
+  }
+
+  ArrayPattern(node: any) {
+    node.elements.forEach(this.visit, this);
+    if (node.decorators) {
+      this.rightHandNodes.push(...node.decorators);
+    }
+    if (node.typeAnnotation) {
+      this.rightHandNodes.push(node.typeAnnotation);
+    }
+  }
+
+  ObjectPattern(node: any) {
+    node.properties.forEach(this.visit, this);
+    if (node.decorators) {
+      this.rightHandNodes.push(...node.decorators);
+    }
+    if (node.typeAnnotation) {
+      this.rightHandNodes.push(node.typeAnnotation);
+    }
+  }
+
+  RestElement(node: any) {
+    super.RestElement(node);
+    if (node.typeAnnotation) {
+      this.rightHandNodes.push(node.typeAnnotation);
+    }
+  }
+}
+
+class Referencer extends OriginalReferencer {
+  constructor(...args: any[]) {
+    super(...args);
+    this.typeMode = false;
+  }
+
+  /**
+   * Override to use PatternVisitor we overrode.
+   * @param {Identifier} node The Identifier node to visit.
+   * @param {Object} [options] The flag to visit right-hand side nodes.
+   * @param {Function} callback The callback function for left-hand side nodes.
+   * @returns {void}
+   */
+  visitPattern(node: any, options: any, callback: any) {
+    if (!node) {
+      return;
+    }
+
+    if (typeof options === 'function') {
+      callback = options;
+      options = { processRightHandNodes: false };
+    }
+
+    const visitor = new PatternVisitor(this.options, node, callback);
+    visitor.visit(node);
+
+    if (options.processRightHandNodes) {
+      visitor.rightHandNodes.forEach(this.visit, this);
+    }
+  }
+
+  /**
+   * Override.
+   * Visit `node.typeParameters` and `node.returnType` additionally to find `typeof` expressions.
+   * @param {FunctionDeclaration|FunctionExpression|ArrowFunctionExpression} node The function node to visit.
+   * @returns {void}
+   */
+  visitFunction(node: any) {
```

Comment:
not sure if you want to do this now, but we have

https://github.com/armano2/typescript-eslint/blob/es-tree-nodes/packages/typescript-estree/src/es-tree-nodes.ts